daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala

```diff
@@ -25,56 +25,92 @@ import scala.collection.immutable.HashMap
 object TransactionCoder {
 
   import ValueCoder.{DecodeCid, EncodeCid, codecContractId}
-  type EncodeNid[Nid] = Nid => String
-  type EncodeVal[Cid <: Value.ContractId] =
-    VersionedValue[Cid] => Either[EncodeError, (ValueVersion, ValueOuterClass.VersionedValue)]
+
+  abstract class EncodeNid[Nid] private[lf] {
+    def asString(id: Nid): String
+  }
+  abstract class DecodeNid[Nid] private[lf] {
+    def fromString(s: String): Either[DecodeError, Nid]
+  }
+
+  val NidEncoder: EncodeNid[Value.NodeId] = new EncodeNid[Value.NodeId] {
+    override def asString(id: Value.NodeId): String = id.index.toString
+  }
+
+  val NidDecoder: DecodeNid[Value.NodeId] = new DecodeNid[Value.NodeId] {
```

Comment:
```suggestion
  val NidDecoder: DecodeNid[Value.NodeId] = s => parseInt(s).fold(...)
```

and so on for the other instances.